### PR TITLE
Fused Multihead Attention for training

### PR DESCRIPTION
**Pull request by fhieber**
Use `torch.nn.functional.multi_head_attention_forward` for self- and encoder-attention during training. Requires some fiddling with the parameter layout of the key-value input projections, as the current Sockeye attention interleaves for faster inference. Attention masks (both for source masking and autoregressive masks need some shape tweaking as requirements for the fused MHA op differ slightly).

Non-interleaved format for joint key-value input projection parameters:
`in_features=hidden, out_features=2*hidden -> Shape: (2*hidden, hidden)`
Interleaved format for joint-key-value input projection stores key and value parameters, grouped by heads:
`Shape: ((num_heads * 2 * hidden_per_head), hidden)`

- Models save and load key-value projection parameters in interleaved format.
- When `model.training == True` key-value projection parameters are put into non-interleaved format for `torch.nn.functional.multi_head_attention_forward`
- When `model.training == False`, i.e. `model.eval()` is called, key-value projection parameters are again converted into interleaved format in place.

This change improves training speed by about 10% but keeps inference speed the same.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by mjdenkowski — APPROVED** [review]
Looks good!
